staging/src/k8s.io/cri-client/pkg/logs/logs.go

```diff
@@ -452,30 +456,76 @@ func isContainerRunning(ctx context.Context, logger *klog.Logger, id string, r i
 	return true, nil
 }
 
-// waitLogs wait for the next log write. It returns two booleans and an error. The first boolean
+const waitDuration time.Duration = time.Millisecond * 150
+
+type waitLog struct {
+	lastCheck      time.Time
+	lastWriteEvent time.Time
+	// There was a create event.
+	recreated bool
+
+	logger         *klog.Logger
+	id             string
+	logName        string
+	w              *fsnotify.Watcher
+	runtimeService internalapi.RuntimeService
+}
+
+func newWaitLog(logger *klog.Logger, id string, logName string, w *fsnotify.Watcher, runtimeService internalapi.RuntimeService) *waitLog {
+	return &waitLog{
+		logger:         logger,
+		id:             id,
+		logName:        logName,
+		w:              w,
+		runtimeService: runtimeService,
+		recreated:      false,
+	}
+}
+
+// wait for the next log write. It returns two booleans and an error. The first boolean
 // indicates whether a new log is found; the second boolean if the log file was recreated;
 // the error is error happens during waiting new logs.
-func waitLogs(ctx context.Context, logger *klog.Logger, id string, logName string, w *fsnotify.Watcher, runtimeService internalapi.RuntimeService) (bool, bool, error) {
-	// no need to wait if the pod is not running
-	if running, err := isContainerRunning(ctx, logger, id, runtimeService); !running {
-		return false, false, err
+func (wl *waitLog) wait(ctx context.Context) (bool, bool, error) {
+	if time.Since(wl.lastCheck) > waitDuration {
+		// no need to wait if the pod is not running
+		if running, err := isContainerRunning(ctx, wl.logger, wl.id, wl.runtimeService); !running {
+			return false, false, err
+		}
+		wl.lastCheck = time.Now()
+	}
+
+	if wl.recreated {
+		wl.recreated = false
+		// Reopen the log file now.
+		return true, true, nil
 	}
+
 	errRetry := 5
 	for {
 		select {
 		case <-ctx.Done():
 			return false, false, fmt.Errorf("context cancelled")
-		case e := <-w.Events:
+		case e := <-wl.w.Events:
 			switch e.Op {
-			case fsnotify.Write, fsnotify.Rename, fsnotify.Remove, fsnotify.Chmod:
+			case fsnotify.Write:
+				if time.Since(wl.lastWriteEvent) > waitDuration {
+					wl.lastWriteEvent = time.Now()
+					return true, false, nil
+				}
```

Comment:
Can you explain how it's guaranteed that even if we skip some write events, eventually all file content will be fully read, especially in the case when file is rotated?

Reply:
Thank you for the explanations!
What happens if in the case 1 file is deleted while we're waiting and we skipped a lot of events. Will the log content that was produced while we're waiting be lost?

So, in case of rotation, old file will still be read even though it's renamed?

Are these cases covered by the tests?

Reply:
As long as a file descriptor remains open, the system will preserve the file from deletion. The application may continue reading the file through this descriptor until it is closed.

This test covers these case.
https://github.com/kubernetes/kubernetes/blob/da61f24acfb54bb8d382d2875ade8fca6af02f79/staging/src/k8s.io/cri-client/pkg/logs/logs_test.go#L215